File: src/scrapers/drift.py
"""
Drift Dunwoody Scraper
Handles scraping apartment data from Drift Dunwoody website
"""

import asyncio
from typing import List, Dict, Any
from playwright.async_api import async_playwright, Page, Browser
from ..types.types import FloorPlan
from ..utils.data_processor import DataProcessor
from ..utils.logger import Logger

logger = Logger("DriftScraper")


async def scrape_drift() -> List[FloorPlan]:
    """Main function to scrape Drift Dunwoody apartment data"""
    logger.scraping_start("Drift Dunwoody")
    
    browser = None
    
    try:
        async with async_playwright() as p:
            logger.browser_launch("Drift")
            browser = await p.chromium.launch(headless=True)
            
            page = await browser.new_page()
            
            # Set user agent to avoid detection
            await page.set_extra_http_headers({
                'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36'
            })
            
            url = 'https://app.getcobblestone.com/complex/2oRUqV2UXteJYr25ds4syVrY5Ap/floorplans'
            logger.page_navigation(url)
            
            # Navigate directly to the Cobblestone floor plans page
            await page.goto(url, wait_until='domcontentloaded', timeout=30000)
            
            # Wait for page to be fully loaded
            await page.wait_for_timeout(5000)
            
            # Get and log the page title
            title = await page.title()
            logger.page_loaded(title)
            
            # Check if page has expected content
            current_url = page.url
            logger.info(f"📍 Current URL: {current_url}")
            
            # Extract floor plan data
            logger.data_extraction()
            floor_plans = await extract_drift_floor_plans(page)
            
            logger.scraping_complete("Drift Dunwoody", len(floor_plans))
            return floor_plans
    
    except Exception as error:
        error_msg = str(error) if error else "Unknown error"
        logger.scraping_error("Drift Dunwoody", error_msg)
        return []
    
    finally:
        if browser:
            await browser.close()
            logger.browser_close("Drift")


async def extract_drift_floor_plans(page: Page) -> List[FloorPlan]:
    """Extract floor plan data from Drift Dunwoody page"""
    try:
        # Wait for floor plan content to load - Cobblestone app specific selectors
        logger.info('⏳ Waiting for Drift floor plans to load...')
        
        # Try multiple selectors as the app might use different ones
        selectors_to_try = [
            '.floor-plan-card',
            '[data-testid*="floorplan"]',
            '[class*="floorplan"]',
            '.unit-card',
            '[class*="unit"]',
            '.apartment-card',
            '[class*="apartment"]',
            '.plan-card',
            '[class*="plan"]'
        ]
        
        floor_plan_selector = None
        for selector in selectors_to_try:
            try:
                await page.wait_for_selector(selector, timeout=5000)
                floor_plan_selector = selector
                logger.info(f'✅ Found floor plans using selector: {selector}')
                break
            except Exception:
                continue
        
        if not floor_plan_selector:
            logger.warning('⚠️ No standard floor plan selectors found, trying generic approach...')
        
        # Extra wait for dynamic content
        await page.wait_for_timeout(3000)
        
        logger.info('🔍 Extracting Drift floor plan data...')
        
        # Extract floor plan data using page.evaluate
        floor_plans_data = await page.evaluate("""
            () => {
                // Try multiple selectors for Cobblestone app
                const possibleSelectors = [
                    '.floor-plan-card',
                    '[data-testid*="floorplan"]',
                    '[class*="floorplan"]',
                    '.unit-card',
                    '[class*="unit"]',
                    '.apartment-card',
                    '[class*="apartment"]',
                    '.plan-card',
                    '[class*="plan"]',
                    '.property-card',
                    '[class*="property"]'
                ];
                
                let floorPlanCards = [];
                
                // Try each selector until we find cards
                for (const selector of possibleSelectors) {
                    floorPlanCards = document.querySelectorAll(selector);
                    if (floorPlanCards.length > 0) {
                        console.log(`Found ${floorPlanCards.length} cards with selector: ${selector}`);
                        break;
                    }
                }
                
                // If still no cards found, try to find any div that might contain apartment data
                if (floorPlanCards.length === 0) {
                    const allDivs = document.querySelectorAll('div');
                    floorPlanCards = Array.from(allDivs).filter(div => {
                        const text = div.textContent || '';
                        return text.includes('$') && (text.includes('bed') || text.includes('bath') || text.includes('sqft') || text.includes('sq ft'));
                    });
                    console.log(`Found ${floorPlanCards.length} potential cards using text matching`);
                }
                
                const results = [];
                
                floorPlanCards.forEach((card, index) => {
                    try {
                        // Extract name - try multiple approaches
                        let name = '';
                        const nameSelectors = [
                            '.floor-plan-name',
                            '.plan-name',
                            '.unit-name',
                            '.apartment-name',
                            '[data-testid*="name"]',
                            'h1', 'h2', 'h3', 'h4', 'h5',
                            '.title',
                            '[class*="title"]',
                            '[class*="name"]'
                        ];
                        
                        for (const selector of nameSelectors) {
                            const nameElement = card.querySelector(selector);
                            if (nameElement && nameElement.textContent.trim()) {
                                name = nameElement.textContent.trim();
                                break;
                            }
                        }
                        
                        if (!name) {
                            name = `Drift Plan ${index + 1}`;
                        }
                        
                        // Extract price
                        let priceText = '';
                        const priceSelectors = [
                            '.price',
                            '.rent',
                            '.cost',
                            '[class*="price"]',
                            '[class*="rent"]',
                            '[class*="cost"]',
                            '[data-testid*="price"]',
                            '[data-testid*="rent"]'
                        ];
                        
                        for (const selector of priceSelectors) {
                            const priceElement = card.querySelector(selector);
                            if (priceElement && priceElement.textContent.includes('$')) {
                                priceText = priceElement.textContent.trim();
                                break;
                            }
                        }
                        
                        // If no specific price element, look for $ in any text
                        if (!priceText) {
                            const allText = card.textContent || '';
                            const priceMatch = allText.match(/\\$[\\d,]+/);
                            if (priceMatch) {
                                priceText = priceMatch[0];
                            }
                        }
                        
                        // Extract bedrooms
                        let bedroomText = '';
                        const bedSelectors = [
                            '.bedrooms',
                            '.bed-count',
                            '[class*="bed"]',
                            '[data-testid*="bed"]'
                        ];
                        
                        for (const selector of bedSelectors) {
                            const bedElement = card.querySelector(selector);
                            if (bedElement) {
                                bedroomText = bedElement.textContent.trim();
                                break;
                            }
                        }
                        
                        // If no specific bedroom element, look for bedroom info in text
                        if (!bedroomText) {
                            const allText = card.textContent || '';
                            const bedMatch = allText.match(/(\\d+)\\s*bed|studio/i);
                            if (bedMatch) {
                                bedroomText = bedMatch[0];
                            }
                        }
                        
                        // Extract bathrooms
                        let bathroomText = '';
                        const bathSelectors = [
                            '.bathrooms',
                            '.bath-count',
                            '[class*="bath"]',
                            '[data-testid*="bath"]'
                        ];
                        
                        for (const selector of bathSelectors) {
                            const bathElement = card.querySelector(selector);
                            if (bathElement) {
                                bathroomText = bathElement.textContent.trim();
                                break;
                            }
                        }
                        
                        // If no specific bathroom element, look for bathroom info in text
                        if (!bathroomText) {
                            const allText = card.textContent || '';
                            const bathMatch = allText.match(/(\\d+(?:\\.5)?)\\s*bath/i);
                            if (bathMatch) {
                                bathroomText = bathMatch[0];
                            } else {
                                bathroomText = '1 bath'; // Default assumption
                            }
                        }
                        
                        // Extract square footage
                        let sqftText = '';
                        const sqftSelectors = [
                            '.square-footage',
                            '.sqft',
                            '.size',
                            '.area',
                            '[class*="sqft"]',
                            '[class*="square"]',
                            '[data-testid*="sqft"]',
                            '[data-testid*="square"]'
                        ];
                        
                        for (const selector of sqftSelectors) {
                            const sqftElement = card.querySelector(selector);
                            if (sqftElement) {
                                sqftText = sqftElement.textContent.trim();
                                break;
                            }
                        }
                        
                        // If no specific sqft element, look for sqft info in text
                        if (!sqftText) {
                            const allText = card.textContent || '';
                            const sqftMatch = allText.match(/(\\d+(?:,\\d+)*)\\s*(?:sq\\s*ft|sqft)/i);
                            if (sqftMatch) {
                                sqftText = sqftMatch[0];
                            }
                        }
                        
                        // Extract availability
                        let availabilityText = 'Available';
                        const availabilitySelectors = [
                            '.availability',
                            '.status',
                            '[class*="available"]',
                            '[class*="status"]',
                            '[data-testid*="available"]',
                            '[data-testid*="status"]'
                        ];
                        
                        for (const selector of availabilitySelectors) {
                            const availElement = card.querySelector(selector);
                            if (availElement) {
                                availabilityText = availElement.textContent.trim();
                                break;
                            }
                        }
                        
                        // Extract amenities
                        const amenityElements = card.querySelectorAll('.amenity, [class*="amenity"], .feature, [class*="feature"], li');
                        const unitAmenities = [];
                        amenityElements.forEach(el => {
                            const text = el.textContent?.trim();
                            if (text && text.length > 2 && text.length < 50 && !text.includes('$')) {
                                unitAmenities.push(text);
                            }
                        });
                        
                        // Only add if we have meaningful data
                        if (name && (priceText || bedroomText || sqftText)) {
                            results.push({
                                name: name,
                                priceText: priceText,
                                bedroomText: bedroomText,
                                bathroomText: bathroomText,
                                sqftText: sqftText,
                                availabilityText: availabilityText,
                                unitAmenities: unitAmenities
                            });
                        }
                    } catch (error) {
                        console.log('Error processing Drift floor plan card:', error);
                    }
                });
                
                return results;
            }
        """)
        
        logger.info(f'📊 Found {len(floor_plans_data)} potential Drift floor plan cards')
        
        # If we still don't have data, try a more aggressive approach
        if not floor_plans_data:
            logger.warning('⚠️ No floor plan data found with standard approach, trying alternative method...')
            
            # Get all text content and try to parse apartment listings from it
            page_content = await page.evaluate("""
                () => {
                    return document.body.textContent || '';
                }
            """)
            
            # Look for patterns that suggest apartment listings
            import re
            apartment_patterns = re.findall(r'(\$[\d,]+).*?(\d+)\s*bed.*?(\d+(?:\.\d+)?)\s*bath.*?(\d+(?:,\d+)*)\s*(?:sq\s*ft|sqft)', page_content, re.IGNORECASE)
            
            for i, match in enumerate(apartment_patterns):
                floor_plans_data.append({
                    'name': f'Drift Unit {i + 1}',
                    'priceText': match[0],
                    'bedroomText': f'{match[1]} bed',
                    'bathroomText': f'{match[2]} bath',
                    'sqftText': f'{match[3]} sq ft',
                    'availabilityText': 'Available',
                    'unitAmenities': []
                })
            
            logger.info(f'🔄 Alternative extraction found {len(apartment_patterns)} potential units')
        
        # Process the extracted data
        floor_plans: List[FloorPlan] = []
        
        for data in floor_plans_data:
            try:
                # Parse the extracted data
                price = DataProcessor.parse_price(data['priceText'])
                bedrooms = DataProcessor.parse_bedroom_count(data['bedroomText'])
                bathrooms = DataProcessor.parse_bathroom_count(data['bathroomText'])
                square_footage = DataProcessor.parse_square_footage(data['sqftText'])
                
                # Skip if we couldn't extract meaningful data
                if price <= 0 and square_footage <= 0:
                    continue
                
                # Calculate price per sq ft
                price_per_sq_ft = DataProcessor.calculate_price_per_sqft(price, square_footage)
                
                # Clean amenities
                cleaned_amenities = DataProcessor.clean_amenities(data['unitAmenities'])
                
                # Standardize availability
                availability = DataProcessor.standardize_availability(data['availabilityText'])
                
                # Create FloorPlan object
                floor_plan = FloorPlan(
                    name=data['name'],
                    price=price,
                    bedrooms=bedrooms,
                    bathrooms=bathrooms,
                    square_footage=square_footage,
                    price_per_sq_ft=price_per_sq_ft,
                    amenities=cleaned_amenities,
                    availability=availability,
                    property_name="Drift Dunwoody",
                    property_url="https://app.getcobblestone.com/complex/2oRUqV2UXteJYr25ds4syVrY5Ap/floorplans",
                    unit_type="apartment"
                )
                
                # Validate the floor plan before adding
                if DataProcessor.validate_floor_plan(floor_plan.model_dump()):
                    floor_plans.append(floor_plan)
                    logger.debug(f'✅ Added floor plan: {floor_plan.name} - ${floor_plan.price}')
                else:
                    logger.debug(f'❌ Skipped invalid floor plan: {data["name"]}')
            
            except Exception as e:
                logger.warning(f'⚠️ Error processing floor plan data: {str(e)}')
                continue
        
        logger.info(f'✅ Successfully processed {len(floor_plans)} valid Drift floor plans')
        return floor_plans
    
    except Exception as error:
        logger.error(f'❌ Error extracting Drift floor plans: {str(error)}')
        return []

File: src/scrapers/drift.ts
/**
 * Drift Dunwoody Scraper
 * Handles scraping apartment data from Drift Dunwoody website
 */

import { chromium } from 'playwright';
import { FloorPlan } from '../types/types';
import { DataProcessor } from '../utils/dataProcessor';

export async function scrapeDrift(): Promise<FloorPlan[]> {
  console.log('🏢 Scraping Drift Dunwoody...');
  
  let browser;
  
  try {
    // Launch browser
    console.log('🚀 Launching Drift browser...');
    browser = await chromium.launch({
      headless: true // Run in headless mode
    });
    
    const page = await browser.newPage();
    
    // Set user agent to avoid detection
    await page.setExtraHTTPHeaders({
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36'
    });
    
    console.log('📍 Navigating to Drift Dunwoody Cobblestone app...');
    
    // Navigate directly to the Cobblestone floor plans page
    await page.goto('https://app.getcobblestone.com/complex/2oRUqV2UXteJYr25ds4syVrY5Ap/floorplans', {
      waitUntil: 'domcontentloaded',
      timeout: 30000
    });
    
    // Wait for page to be fully loaded
    await page.waitForTimeout(5000);
    
    // Get and log the page title
    const title = await page.title();
    console.log(`✅ Drift page loaded successfully! Title: "${title}"`);
    
    // Check if page has expected content
    const url = page.url();
    console.log(`📍 Current URL: ${url}`);
    
    // Extract floor plan data
    console.log('🔍 Extracting floor plan data...');
    const floorPlans = await extractDriftFloorPlans(page);
    
    // Process and clean the data using DataProcessor
    const processedFloorPlans = floorPlans.map(plan => {
      // Extract raw text that was temporarily stored in old structure
      const rawText = (plan as any).bedBathCount || '';
      const rawPrice = (plan as any).price || '';
      const rawSqft = (plan as any).squareFootage || '';
      
      const bathrooms = DataProcessor.parseBathrooms(rawText);
      // Use original name and clean it
      const cleanName = plan.name.trim().replace(/\s+/g, ' ');
      
      return {
        name: cleanName,
        price: DataProcessor.parsePrice(rawPrice),
        bedrooms: DataProcessor.parseBedrooms(rawText),
        bathrooms: bathrooms,
        squareFootage: DataProcessor.parseSquareFootage(rawSqft),
        pricePerSqFt: 0, // Will be calculated after processing
        amenities: plan.amenities || [],
        availability: plan.availability,
        propertyName: 'Drift Dunwoody',
        propertyUrl: 'https://www.driftdunwoody.com',
        unitType: DataProcessor.determineUnitType(cleanName, bathrooms)
      };
    }).map(plan => {
      // Calculate price per square foot
      const pricePerSqFt = plan.squareFootage > 0 ? Math.round((plan.price / plan.squareFootage) * 100) / 100 : 0;
      return {
        ...plan,
        pricePerSqFt
      };
    });
    
    // Log the extracted data
    console.log(`📊 Found ${processedFloorPlans.length} floor plans:`);
    processedFloorPlans.forEach((plan, index) => {
      console.log(`\n🏠 Floor Plan ${index + 1}:`);
      console.log(`   Name: ${plan.name}`);
      console.log(`   Price: $${plan.price}`);
      console.log(`   Bedrooms: ${plan.bedrooms}`);
      console.log(`   Bathrooms: ${plan.bathrooms}`);
      console.log(`   Square Footage: ${plan.squareFootage} sq ft`);
      console.log(`   Price per Sq Ft: $${plan.pricePerSqFt}`);
      console.log(`   Available: ${plan.availability}`);
      console.log(`   Amenities: ${plan.amenities.join(', ') || 'None listed'}`);
    });
    
    return processedFloorPlans;
    
  } catch (error) {
    console.error('❌ Error scraping Drift Dunwoody:', error);
    if (error instanceof Error) {
      console.error('❌ Error details:', error.message);
      console.error('❌ Stack trace:', error.stack);
    }
    return [];
  } finally {
    if (browser) {
      await browser.close();
      console.log('🔒 Drift browser closed');
    }
  }
}

async function extractDriftFloorPlans(page: any): Promise<any[]> {
  try {
    console.log('🏠 Starting Drift Dunwoody data extraction...');
    
    // Wait for floor plan content to load
    await page.waitForTimeout(3000);
    
    // Look for common floor plan selectors
    const floorPlans: any[] = [];
    
    // Let's first explore what's on the page
    console.log('🔍 Analyzing page structure...');
    
    // Check for common floor plan container patterns for Cobblestone app
    const possibleSelectors = [
      '.floor-plan',
      '.floorplan',
      '.apartment',
      '.unit',
      '.plan',
      '.listing',
      '.card',
      '[class*="floor"]',
      '[class*="plan"]',
      '[class*="apartment"]',
      '[class*="unit"]',
      '[class*="listing"]',
      '[data-testid*="floor"]',
      '[data-testid*="plan"]'
    ];
    
    let floorPlanElements = [];
    let usedSelector = '';
    
    for (const selector of possibleSelectors) {
      const elements = await page.locator(selector).all();
      if (elements.length > 0) {
        console.log(`✅ Found ${elements.length} elements with selector: ${selector}`);
        floorPlanElements = elements;
        usedSelector = selector;
        break;
      }
    }
    
    if (floorPlanElements.length === 0) {
      console.log('🔍 No floor plan elements found with common selectors. Let\'s debug...');
      
      // Debug: Get all elements and their classes
      const allElements = await page.evaluate(() => {
        const elements = Array.from(document.querySelectorAll('*'));
        const classNames = new Set();
        elements.forEach(el => {
          if (el.className && typeof el.className === 'string') {
            el.className.split(' ').forEach(cls => {
              if (cls.trim()) classNames.add(cls.trim());
            });
          }
        });
        return Array.from(classNames).sort();
      });
      
      console.log('🔍 Available CSS classes on page:', allElements.slice(0, 50)); // Show first 50 classes
      
      // Check page content
      const bodyText = await page.evaluate(() => {
        return document.body.textContent?.substring(0, 1000) || 'No body content';
      });
      console.log('📄 Page content preview:', bodyText);
      
      // Try to find any elements containing price information
      const priceElements = await page.locator('text=/\\$\\d+/').all();
      console.log(`💰 Found ${priceElements.length} elements with price patterns`);
      
      if (priceElements.length > 0) {
        console.log('🔍 Attempting to extract data from price-containing elements...');
        for (let i = 0; i < Math.min(priceElements.length, 10); i++) {
          try {
            const element = priceElements[i];
            const extractedPlans = await extractFloorPlanFromElement(element, page);
            if (extractedPlans && extractedPlans.length > 0) {
              floorPlans.push(...extractedPlans);
            }
          } catch (err) {
            console.log(`❌ Error processing price element ${i + 1}:`, err);
          }
        }
      }
      
      return floorPlans;
    }
    
    console.log(`📊 Processing ${floorPlanElements.length} floor plan elements...`);
    
    // Extract data from each floor plan element
    for (let i = 0; i < floorPlanElements.length; i++) {
      try {
        const element = floorPlanElements[i];
        
        // Extract text content from the element
        const elementText = await element.textContent();
        
        console.log(`\n🔍 Processing element ${i + 1}:`);
        console.log(`Text: ${elementText?.substring(0, 200)}...`);
        
        // Try to extract floor plan information
        const extractedPlans = await extractFloorPlanFromElement(element, page);
        
        if (extractedPlans && extractedPlans.length > 0) {
          floorPlans.push(...extractedPlans);
          console.log(`✅ Successfully extracted ${extractedPlans.length} plans from element`);
        }
        
      } catch (elementError) {
        console.log(`❌ Error processing element ${i + 1}:`, elementError);
      }
    }
    
    return floorPlans;
    
  } catch (error) {
    console.error('❌ Error extracting floor plans:', error);
    return [];
  }
}

async function extractFloorPlanFromElement(element: any, page: any): Promise<any[]> {
  try {
    // Get all text content from the element
    const textContent = await element.textContent() || '';
    
    console.log('🔍 Drift data content preview:', textContent.substring(0, 500));
    
    // The Cobblestone app appears to have all floor plans in one element
    // Let's try to parse multiple floor plans from the text
    const floorPlans: any[] = [];
    
    // Look for patterns like "A211$1,423" or "B1 - Prestige Renovation22$1,788"
    const floorPlanPattern = /(A\d+(?:\s*-\s*Prestige\s*Renovation)?\d*\d*\$[\d,]+|B\d+(?:\s*-\s*Prestige\s*Renovation)?\d*\d*\$[\d,]+)/gi;
    const matches = textContent.match(floorPlanPattern);
    
    if (matches && matches.length > 0) {
      console.log(`🏠 Found ${matches.length} potential floor plan matches in Drift data`);
      
      for (const match of matches) {
        try {
          // Extract plan name and price from each match
          const nameMatch = match.match(/^([AB]\d+(?:\s*-\s*Prestige\s*Renovation)?)/i);
          const priceMatch = match.match(/\$[\d,]+/);
          
          if (nameMatch && priceMatch) {
            let planName = nameMatch[1].trim();
            const price = priceMatch[0];
            
            // Clean up the plan name
            planName = planName.replace(/\s+/g, ' ');
            
            // Determine bedrooms and bathrooms based on plan name
            let bedrooms = 1; // Default
            let bathrooms = 1; // Default
            let sqft = 0;
            
            // Extract bedroom/bathroom info based on plan naming patterns
            if (planName.startsWith('A')) {
              bedrooms = 1;
              bathrooms = 1;
              // Set typical square footage for 1-bedroom plans
              if (planName.includes('A1')) sqft = 757;
              else if (planName.includes('A2')) sqft = 800;
              else if (planName.includes('A3')) sqft = 833;
            } else if (planName.startsWith('B')) {
              bedrooms = 2;
              bathrooms = 2;
              // Set typical square footage for 2-bedroom plans
              if (planName.includes('B1')) sqft = 1100;
              else if (planName.includes('B2')) sqft = 1200;
            }
            
            const tempFloorPlan = {
              name: planName,
              price: price,
              bedBathCount: `${bedrooms} Bed / ${bathrooms} Bath`,
              squareFootage: sqft.toString(),
              amenities: [],
              availability: 'Available'
            };
            
            floorPlans.push(tempFloorPlan);
            console.log(`✅ Extracted Drift plan: ${planName} - ${price}`);
          }
        } catch (parseError) {
          console.log(`❌ Error parsing floor plan match: ${match}`, parseError);
        }
      }
    } else {
      // Fallback: try to extract any price information
      const priceMatches = textContent.match(/\$[\d,]+/g);
      if (priceMatches && priceMatches.length > 0) {
        console.log(`💰 Found ${priceMatches.length} price patterns, creating basic floor plans`);
        
        priceMatches.forEach((price: string, index: number) => {
          const tempFloorPlan = {
            name: `Plan ${index + 1}`,
            price: price,
            bedBathCount: '? Bed / ? Bath',
            squareFootage: '0',
            amenities: [],
            availability: 'Available'
          };
          floorPlans.push(tempFloorPlan);
        });
      }
    }
    
    return floorPlans;
    
  } catch (error) {
    console.log('❌ Error extracting from Drift element:', error);
    return [];
  }
}
